cli/tests/test_help_command.rs

```diff
@@ -0,0 +1,57 @@
+// Copyright 2024 The Jujutsu Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::common::TestEnvironment;
+
+#[test]
+fn test_help() {
+    let test_env = TestEnvironment::default();
+
+    let help_cmd_stdout = test_env.jj_cmd_success(test_env.env_root(), &["help"]);
+    // The help command output should be equal to the long --help flag
+    let help_flag_stdout = test_env.jj_cmd_success(test_env.env_root(), &["--help"]);
+    assert_eq!(help_cmd_stdout, help_flag_stdout);
+
+    // Help command should work with commands
+    let help_cmd_stdout = test_env.jj_cmd_success(test_env.env_root(), &["help", "log"]);
+    let help_flag_stdout = test_env.jj_cmd_success(test_env.env_root(), &["log", "--help"]);
+    assert_eq!(help_cmd_stdout, help_flag_stdout);
+
+    // Help command should work with subcommands
+    let help_cmd_stdout =
+        test_env.jj_cmd_success(test_env.env_root(), &["help", "workspace", "root"]);
+    let help_flag_stdout =
+        test_env.jj_cmd_success(test_env.env_root(), &["workspace", "root", "--help"]);
+    assert_eq!(help_cmd_stdout, help_flag_stdout);
```

Comment:
For me, the `--help` flag should work like a global flag but `jj --help nonexistent` should not work, as you are asking help for a nonexistent subcommand.

Reply:
I agree. The only reason I said "probably" is that it's currently supported. I'm not sure why it is (deliberate behavior by clap? bug in how we use it?), but that doesn't matter.

Reply:
Yeah, I checked that it works after I read your comment. Bug or not, for me, it's an estrange behavior. Is there something to do here?

Reply:
I think it would be good to have tests of the `--help` flag in various positions. And it would be good to have tests with non-existent commands too, both `jj help nonexistent` and `jj nonexistent --help`. I assume the latter will be handled by clap, so just check check that the output is the same, as with the existing tests. Thanks.

Reply:
Hum, actually the current behavior is not what I think is the ideal.

`jj workspace root --help` and `jj workspace --help root` give different outputs, the first one gives the help of `jj workspace root` and the latter gives the help for `jj workspace`. For me, both should give the help for `jj workspace root`. What you guys think?

Reply:
iirc, that's because clap ignores arguments after `--help`. I don't know the exact reason, but it might be useful if `--help` can be inserted in failing command arguments without removing bad parts.

Reply:
Maybe @epage can tell us

Reply:
imo the expectation is that `--help` is an argument for the command that is being applied to.

Some more concrete thoughts

The current approach of clap's allows you to edit an existing command and add `--help` anywhere to get details on that part of it.

`--help` bails out immediately.  If we treated it as a global flag that runs against whatever subcommands are specified, then any kind of validation error could prevent you from seeing `--help`.  We could go into an `ignore_errors` mode but that is dicey.

Reply:
Do you mean that something like `jj log --help nonexistent` or `jj log --help --nonexistent` would prevent help for `log` from being displayed, and instead resulting in the top-level help being displayed? I actually think it sounds reasonable to do that. `jj help nonexistent` and `jj nonexistent --help` currently error out. I would actually prefer for `jj --help nonexistent` to error out in the same way.

Reply:
I'm not just talking about `jj --help non-existent` but any other parse errors.